lib/Monad_WP/OptionMonad.thy

```diff
@@ -88,6 +112,18 @@ lemma opt_pred_disj:
   "((P1 |< hp) p \<or> (P2 |< hp) p) = (((P1 or P2) |< hp) p)"
   by (fastforce simp: pred_disj_def split: option.splits)
 
+lemma opt_predD:
+  "(P |< proj) x \<Longrightarrow> \<exists>y. proj x = Some y \<and> P y"
+  by (clarsimp split: option.splits)
+
+lemma opt_pred_unfold:
```

Comment:
This should be called `opt_pred_assoc`, I think. If there is no other reason against, it should be oriented the other way around, and potentially be `[simp]` so that parentheses are automatically removed (same as for `+` etc).

Reply:
Hm, this is a bit more subtle than a standard `assoc`, maybe `opt_pred_map_assoc` to express that two operators are involved here. Maybe even `opt_pred_map_distrib`.

Reply:
It is actually the one below that I had the need to write, and this is a full `opt_map` version of it.

The idea for having this and the one below is to tweak what sits lhs and rhs of the top level `opt_pred` (`|<`) so that the rhs is a particular projection you need. Then we can combine it with another `opt_pred` statement via `opt_pred_conj`, etc., and it would be nice if we can do some reasoning inside the lhs of `opt_pred`.
So, I'm roughly aiming to have something like this:
`(P |< f |> g) and (P' |< f) = ((P |< g and P') |< f)`
`(P |< f ||> proj) and (P' |< f) = ((P o proj and P') |< f)`

Reply:
Those look both good. I think you could prove them still with the rule oriented in the other direction (this one, the one below is fine).  Maybe not, depending on how it is applied. Happy to leave that to your experience with the rule.